### ISSUE TITLE
Clang in Cygwin: -o /dev/stdout may confuse output redirection

### ISSUE THREAD

**Issue by pmor13** [llvm/llvm-project#54086]
Scenario:
```
$ clang t455.c -S -o /dev/stdout
        .text
        <asm code>

# introduce the error

$ clang t455.c -S -o /dev/stdout
t455.c:26:1: error: unknown type name 'x'
x
^
t455.c:26:2: error: expected identifier or '('
x
 ^
2 errors generated.

# fix the error

$ clang t455.c -S -o /dev/stdout
# nothing is printed

$ clang --version
clang version 8.0.1 (tags/RELEASE_801/final)
```
Why `clang t455.c -S -o /dev/stdout` leads to nothing is printed?

Note: GCC on Cygwin does not have this issue.

**Comment by EugeneZelenko** [comment]
Clang 8 is very old version. Does same problem exist in `main` or Clang 13?

**Comment by pmor13** [comment]
@EugeneZelenko Yes, the problem was initially discovered on Clang 13.0.0 on Cygwin.

**Comment by pmor13** [comment]
@EugeneZelenko Also: `echo "Hello World" > /dev/stdout` leads to nothing is printed.

**Comment by MaskRay** [comment]
Please provide `t455.c`: perhaps edit the comment to include `echo 'int foo() { return x; }' > t455.c`

Did you mean

> `-o /dev/stdout may confuse output redirection`

`-o /dev/stdout` may accidentally redirect stderr in Cygwin? This is not a problem on Linux.

**Comment by pmor13** [comment]
@EugeneZelenko @MaskRay If that helps:
```
$ ls -l /dev/stdout
-rw-r--r-- 1 xxx None 12 Feb 26 00:21 /dev/stdout

$ ls -l /dev/stderr
lrwxrwxrwx 1 xxx None 15 Oct 18  2019 /dev/stderr -> /proc/self/fd/2
```

**Comment by efriedma-quic** [comment]
When clang errors out, it `unlink()`s the output file.  For normal files, that's an expected part of how the compiler works.  For /dev/stdout, I guess that isn't expected, but it would normally fail with EACCES, assuming you aren't running clang as root.  I guess maybe on Windows, it's more likely to end up running cygwin as root?

There's actually a bit of a weird inconsistency with clang's output handling when symlinks are involved.  If "-o" specifies a symlink, when clang successfully compiles something, it writes to the destination of the symlink, but when it fails to compile, it erases the symlink itself.  Other tools like gcc apparently do the same thing.  There might be some wiggle room for adjusting our behavior here a bit, though.

**Comment by pmor13** [comment]
@efriedma-quic I run Cygwin as a user. Extra question: even after reboot the problem persists:
```
$ file /dev/std*
/dev/stderr: symbolic link to /proc/self/fd/2
/dev/stdin:  symbolic link to /proc/self/fd/0
/dev/stdout: ASCII text
```
Any idea how to recover `/dev/stdout`?
